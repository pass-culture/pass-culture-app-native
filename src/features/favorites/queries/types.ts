import { FavoriteResponse } from 'api/gen'

export interface FavoriteMutationContext {
  previousFavorites: Array<FavoriteResponse>
}
