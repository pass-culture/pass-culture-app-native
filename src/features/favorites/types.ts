export type FavoriteSortBy = 'RECENTLY_ADDED' | 'ASCENDING_PRICE' | 'AROUND_ME'
