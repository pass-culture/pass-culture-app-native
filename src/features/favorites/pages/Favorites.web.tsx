import { Link } from '@react-navigation/native'
import React from 'react'
import { Text } from 'react-native'
import styled from 'styled-components/native'

// TODO WEB : delete this module and implement web version of this screen
export const Favorites = () => (
  <Page>
    <Text>Favorites</Text>
    <Link to={'/abtesting'}>
      <Text>Go to ABTestingPOC</Text>
    </Link>
    <Link to={'/search'}>
      <Text>Go to Search</Text>
    </Link>
  </Page>
)

const Page = styled.View({
  flex: 1,
  alignItems: 'center',
  justifyContent: 'center',
})
