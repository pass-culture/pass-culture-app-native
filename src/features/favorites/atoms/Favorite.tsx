import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import React, { useMemo, useRef, useState } from 'react'
import { Animated, LayoutChangeEvent } from 'react-native'
import { useQueryClient } from 'react-query'
import styled, { useTheme } from 'styled-components/native'

import { FavoriteOfferResponse, FavoriteResponse, UserProfileResponse } from 'api/gen'
import { useRemoveFavorite } from 'features/favorites/pages/useFavorites'
import { UseNavigationType } from 'features/navigation/RootNavigator'
import { mergeOfferData } from 'features/offer/atoms/OfferTile'
import { OfferImage } from 'features/search/atoms/OfferImage'
import { analytics } from 'libs/analytics'
import { useDistance } from 'libs/geolocation/hooks/useDistance'
import { formatToFrenchDate, getFavoriteDisplayPrice } from 'libs/parsers'
import { QueryKeys } from 'libs/queryKeys'
import { GLOBAL_STALE_TIME } from 'libs/react-query/queryClient'
import { useSearchGroupLabel, useSubcategory } from 'libs/subcategories'
import { ButtonSecondary } from 'ui/components/buttons/ButtonSecondary'
import { SNACK_BAR_TIME_OUT, useSnackBarContext } from 'ui/components/snackBar/SnackBarContext'
import { getSpacing, Spacer, Typo } from 'ui/theme'
import { getHeadingAttrs } from 'ui/theme/typography'
import { Link } from 'ui/web/link/Link'

import { BookingButton } from './BookingButton'

interface Props {
  favorite: FavoriteResponse
  onInAppBooking: (bookedOffer: FavoriteOfferResponse) => void
  user: UserProfileResponse
}

export const Favorite: React.FC<Props> = (props) => {
  const { offer } = props.favorite
  const theme = useTheme()
  const [height, setHeight] = useState<number | undefined>(undefined)
  const animatedOpacity = useRef(new Animated.Value(1)).current
  const animatedCollapse = useRef(new Animated.Value(1)).current
  const { navigate } = useNavigation<UseNavigationType>()
  const queryClient = useQueryClient()
  const distanceToOffer = useDistance({
    lat: offer.coordinates?.latitude,
    lng: offer.coordinates?.longitude,
  })
  const { showErrorSnackBar } = useSnackBarContext()
  const { categoryId, searchGroupName } = useSubcategory(offer.subcategoryId)
  const searchGroupLabel = useSearchGroupLabel(searchGroupName)

  const { mutate: removeFavorite, isLoading } = useRemoveFavorite({
    onError: () => {
      showErrorSnackBar({
        message: t`L'offre n'a pas été retirée de tes favoris`,
        timeout: SNACK_BAR_TIME_OUT,
      })
    },
  })

  const formattedDate = useMemo(() => {
    if (offer.date) {
      return formatToFrenchDate(new Date(offer.date))
    }
    if (offer.startDate) {
      return t({
        id: 'starting from date',
        values: { date: formatToFrenchDate(new Date(offer.startDate)) },
        message: 'Dès le {date}',
      })
    }
    return null
  }, [offer])

  function handlePressOffer() {
    // We pre-populate the query-cache with the data from the search result for a smooth transition
    if (!offer.id) return
    queryClient.setQueryData(
      [QueryKeys.OFFER, offer.id],
      mergeOfferData({
        ...offer,
        categoryId,
        thumbUrl: offer.image?.url,
        name: offer.name,
        offerId: offer.id,
      }),
      {
        // Make sure the data is stale, so that it is considered as a placeholder
        updatedAt: Date.now() - (GLOBAL_STALE_TIME + 1),
      }
    )
    analytics.logConsultOffer({ offerId: offer.id, from: 'favorites' })
    navigate('Offer', { id: offer.id, from: 'favorites' })
  }

  function onRemove() {
    Animated.parallel([
      Animated.timing(animatedOpacity, {
        toValue: 0,
        delay: 0,
        duration: 300,
        useNativeDriver: false,
      }),
      Animated.timing(animatedCollapse, {
        toValue: 0,
        delay: 100,
        duration: 300,
        useNativeDriver: false,
      }),
    ]).start(() => {
      removeFavorite(props.favorite.id)
    })
  }

  function onLayout(event: LayoutChangeEvent) {
    const { height: newHeight } = event.nativeEvent.layout
    if (!height) {
      setHeight(newHeight)
    }
  }

  return (
    <Animated.View
      onLayout={onLayout}
      style={{
        opacity: animatedOpacity,
        height: height
          ? animatedCollapse.interpolate({
              inputRange: [0, 1],
              outputRange: [0, height],
            })
          : undefined,
      }}>
      <Container onPress={handlePressOffer} testID="favorite">
        <Link
          to={{ screen: 'Offer', params: { id: offer.id, from: 'favorites' } }}
          accessible={false}>
          <Row>
            <OfferImage imageUrl={offer.image?.url} categoryId={categoryId} />
            <Spacer.Row numberOfSpaces={4} />
            <Column>
              <Row>
                {distanceToOffer ? (
                  <React.Fragment>
                    <Spacer.Flex flex={0.7}>
                      <Name numberOfLines={2}>{offer.name}</Name>
                    </Spacer.Flex>
                    <Spacer.Flex flex={0.3}>
                      <Distance>{distanceToOffer}</Distance>
                    </Spacer.Flex>
                  </React.Fragment>
                ) : (
                  <Name numberOfLines={2}>{offer.name}</Name>
                )}
              </Row>
              <Spacer.Column numberOfSpaces={1} />
              <Body>{searchGroupLabel}</Body>
              {!!formattedDate && <Body>{formattedDate}</Body>}
              <Spacer.Column numberOfSpaces={1} />
              <Typo.Caption>
                {getFavoriteDisplayPrice({ startPrice: offer.startPrice, price: offer.price })}
              </Typo.Caption>
            </Column>
          </Row>
        </Link>
      </Container>
      <ButtonsRow>
        <ButtonContainer>
          <ButtonSecondary
            wording={t`Supprimer`}
            accessibilityLabel={t`Supprimer l'offre de mes favoris`}
            onPress={onRemove}
            buttonHeight="tall"
            disabled={isLoading}
          />
        </ButtonContainer>
        {!theme.isMobileViewport && <Spacer.Flex flex={1 / 30} />}
        <ButtonContainer>
          <BookingButton offer={offer} user={props.user} onInAppBooking={props.onInAppBooking} />
        </ButtonContainer>
      </ButtonsRow>
      <Separator />
    </Animated.View>
  )
}

const imageWidth = getSpacing(16)

const Container = styled.TouchableOpacity.attrs(({ theme }) => ({
  activeOpacity: theme.activeOpacity,
}))({
  marginHorizontal: getSpacing(6),
})

const columnPadding = 4
const columnMargin = 2 * 6

const Column = styled.View(({ theme }) => ({
  width: theme.appContentWidth - getSpacing(columnMargin + columnPadding) - imageWidth,
}))

const Row = styled.View({ flexDirection: 'row', alignItems: 'center' })

const ButtonContainer = styled.View({
  minWidth: getSpacing(30),
  maxWidth: getSpacing(70),
  width: '47%',
})

const ButtonsRow = styled.View(({ theme }) => ({
  flexDirection: 'row',
  justifyContent: theme.isMobileViewport ? 'space-between' : 'center',
  marginTop: getSpacing(6),
  marginHorizontal: getSpacing(6),
}))

const Name = styled(Typo.ButtonText).attrs(() => getHeadingAttrs(3))``

const Distance = styled(Typo.Body)(({ theme }) => ({
  textAlign: 'right',
  color: theme.colors.greyDark,
}))

const Body = styled(Typo.Body)(({ theme }) => ({
  color: theme.colors.greyDark,
}))

const Separator = styled.View(({ theme }) => ({
  height: 2,
  backgroundColor: theme.colors.greyLight,
  marginHorizontal: getSpacing(6),
  marginVertical: getSpacing(4),
}))
