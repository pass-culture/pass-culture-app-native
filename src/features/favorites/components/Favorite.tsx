import React, { useCallback, useMemo, useRef } from 'react'
import { Animated } from 'react-native'
import styled from 'styled-components/native'

import { FavoriteOfferResponse, FavoriteResponse, UserProfileResponse } from 'api/gen'
import { useRemoveFavorite } from 'features/favorites/api'
import { getBookingButtonProperties } from 'features/favorites/helpers/getBookingButtonProperties'
import { getFavoriteDisplayPrice } from 'features/favorites/helpers/getFavoriteDisplayPrice'
import { useFavoriteFormattedDate } from 'features/favorites/helpers/useFavoriteFormattedDate'
import { StepperOrigin } from 'features/navigation/RootNavigator/types'
import { getShareOffer } from 'features/share/helpers/useShareOffer'
import { WebShareModal } from 'features/share/pages/WebShareModal'
import { analytics } from 'libs/analytics'
import { useDistance } from 'libs/geolocation/hooks/useDistance'
import { useSearchGroupLabel, useSubcategory } from 'libs/subcategories'
import { tileAccessibilityLabel, TileContentType } from 'libs/tileAccessibilityLabel'
import { useBookOfferModal } from 'shared/offer/helpers/useBookOfferModal'
import { usePrePopulateOffer } from 'shared/offer/usePrePopulateOffer'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { ButtonSecondary } from 'ui/components/buttons/ButtonSecondary'
import { RoundedButton } from 'ui/components/buttons/RoundedButton'
import { useModal } from 'ui/components/modals/useModal'
import { SNACK_BAR_TIME_OUT, useSnackBarContext } from 'ui/components/snackBar/SnackBarContext'
import { OfferImage } from 'ui/components/tiles/OfferImage'
import { ExternalTouchableLink } from 'ui/components/touchableLink/ExternalTouchableLink'
import { InternalTouchableLink } from 'ui/components/touchableLink/InternalTouchableLink'
import { useElementHeight } from 'ui/hooks/useElementHeight'
import { ExternalSite } from 'ui/svg/icons/ExternalSite'
import { getSpacing, Spacer, Typo } from 'ui/theme'

interface Props {
  favorite: FavoriteResponse
  onInAppBooking: (bookedOffer: FavoriteOfferResponse) => void
  user: UserProfileResponse
}

const SPACER_BETWEEN_IMAGE_AND_CONTENT = 4

export const Favorite: React.FC<Props> = (props) => {
  const { offer } = props.favorite
  const { onLayout, height } = useElementHeight()
  const animatedOpacity = useRef(new Animated.Value(1)).current
  const animatedCollapse = useRef(new Animated.Value(1)).current
  const prePopulateOffer = usePrePopulateOffer()
  const distanceToOffer = useDistance({
    lat: offer.coordinates?.latitude,
    lng: offer.coordinates?.longitude,
  })
  const displayPrice = getFavoriteDisplayPrice({ startPrice: offer.startPrice, price: offer.price })
  const { showErrorSnackBar } = useSnackBarContext()
  const { categoryId, searchGroupName } = useSubcategory(offer.subcategoryId)
  const searchGroupLabel = useSearchGroupLabel(searchGroupName)
  const formattedDate = useFavoriteFormattedDate({ offer })
  const { modalToDisplay, ...buttonProperties } =
    getBookingButtonProperties({
      offer,
      user: props.user,
      onInAppBooking: props.onInAppBooking,
    }) ?? {}

  const { OfferModal: BookOfferModal, showModal: showBookOfferModal } = useBookOfferModal({
    modalToDisplay,
    offerId: offer.id,
    from: StepperOrigin.FAVORITE,
  })

  const { mutate: removeFavorite, isLoading } = useRemoveFavorite({
    onError: () => {
      showErrorSnackBar({
        message: 'L’offre n’a pas été retirée de tes favoris',
        timeout: SNACK_BAR_TIME_OUT,
      })
    },
  })

  const accessibilityLabel = tileAccessibilityLabel(TileContentType.OFFER, {
    ...offer,
    categoryLabel: searchGroupLabel,
    distance: distanceToOffer,
    date: formattedDate,
    price: displayPrice,
  })

  function handlePressOffer() {
    // We pre-populate the query-cache with the data from the search result for a smooth transition
    if (!offer.id) return
    prePopulateOffer({
      ...offer,
      categoryId,
      thumbUrl: offer.image?.url,
      name: offer.name,
      offerId: offer.id,
    })

    analytics.logConsultOffer({ offerId: offer.id, from: 'favorites' })
  }

  function onRemove() {
    Animated.parallel([
      Animated.timing(animatedOpacity, {
        toValue: 0,
        delay: 0,
        duration: 300,
        useNativeDriver: false,
      }),
      Animated.timing(animatedCollapse, {
        toValue: 0,
        delay: 100,
        duration: 300,
        useNativeDriver: false,
      }),
    ]).start(() => {
      removeFavorite(props.favorite.id)
    })
  }

  const animatedViewStyle = {
    opacity: animatedOpacity,
    height: height
      ? animatedCollapse.interpolate({
          inputRange: [0, 1],
          outputRange: [0, height],
        })
      : undefined,
  }

  const {
    visible: shareOfferModalVisible,
    showModal: showShareOfferModal,
    hideModal: hideShareOfferModal,
  } = useModal(false)

  const { share: shareOffer, shareContent } = getShareOffer({
    offerId: offer.id,
    offerName: offer.name,
    venueName: offer.venueName,
    utmMedium: 'favorite',
  })

  const pressShareOffer = useCallback(() => {
    analytics.logShare({ type: 'Offer', from: 'favorites', offer_id: offer.id })
    shareOffer()
    showShareOfferModal()
  }, [offer.id, shareOffer, showShareOfferModal])

  const BookingButton = useMemo(() => {
    const { wording, externalNav, disabled, accessibilityLabel } = buttonProperties
    if (!wording) return
    if (externalNav)
      return (
        <ExternalTouchableLink
          externalNav={externalNav}
          wording={wording}
          accessibilityLabel={accessibilityLabel}
          as={ButtonPrimary}
          icon={ExternalSite}
          buttonHeight="tall"
        />
      )
    const onPressBookingButton = () => {
      buttonProperties?.onPress?.()
      showBookOfferModal()
    }
    return (
      <ButtonPrimary
        wording={wording}
        disabled={disabled}
        onPress={onPressBookingButton}
        buttonHeight="tall"
      />
    )
  }, [buttonProperties, showBookOfferModal])

  return (
    <React.Fragment>
      <Animated.View onLayout={onLayout} style={animatedViewStyle}>
        <Container>
          <StyledTouchableLink
            navigateTo={
              offer.id
                ? { screen: 'Offer', params: { id: offer.id, from: 'favorites' } }
                : undefined
            }
            onBeforeNavigate={handlePressOffer}
            accessibilityLabel={accessibilityLabel}>
            <Row>
              <OfferImage imageUrl={offer.image?.url} categoryId={categoryId} />
              <Spacer.Row numberOfSpaces={SPACER_BETWEEN_IMAGE_AND_CONTENT} />
              <ContentContainer>
                <LeftContent>
                  <Typo.ButtonText numberOfLines={2}>{offer.name}</Typo.ButtonText>
                  <Spacer.Column numberOfSpaces={1} />
                  <Body>{searchGroupLabel}</Body>
                  {!!formattedDate && <Body>{formattedDate}</Body>}
                  {!!displayPrice && (
                    <React.Fragment>
                      <Spacer.Column numberOfSpaces={1} />
                      <Typo.Caption>{displayPrice}</Typo.Caption>
                    </React.Fragment>
                  )}
                </LeftContent>
                <Spacer.Row numberOfSpaces={2} />
                <RightContent>
                  {!!distanceToOffer && <Distance>{distanceToOffer}</Distance>}
                </RightContent>
              </ContentContainer>
            </Row>
          </StyledTouchableLink>
          <ShareContainer>
            <RoundedButton
              iconName="share"
              onPress={pressShareOffer}
              accessibilityLabel={`Partager l’offre ${offer.name}`}
            />
          </ShareContainer>
        </Container>
        <FavoriteButtonsContainer>
          <ButtonContainer>
            <ButtonSecondary
              wording="Supprimer"
              accessibilityLabel={`Supprimer l’offre ${offer.name} de mes favoris`}
              onPress={onRemove}
              buttonHeight="tall"
              disabled={isLoading}
            />
          </ButtonContainer>
          <Spacer.Row numberOfSpaces={5} />
          <ButtonContainer>{BookingButton}</ButtonContainer>
        </FavoriteButtonsContainer>
        <Separator />
      </Animated.View>
      {!!shareContent && (
        <WebShareModal
          visible={shareOfferModalVisible}
          headerTitle="Partager l’offre"
          shareContent={shareContent}
          dismissModal={hideShareOfferModal}
        />
      )}
      {BookOfferModal}
    </React.Fragment>
  )
}

const Container = styled.View({
  position: 'relative',
})

const ShareContainer = styled.View(({ theme }) => ({
  position: 'absolute',
  top: 0,
  right: 0,
  marginRight: theme.contentPage.marginHorizontal,
}))

const LeftContent = styled.View({
  flex: 1,
})

const RightContent = styled.View(({ theme }) => ({
  justifyContent: 'flex-end',
  alignItems: 'flex-end',
  minWidth: theme.buttons.roundedButton.size,
}))

const StyledTouchableLink = styled(InternalTouchableLink)(({ theme }) => ({
  marginHorizontal: theme.contentPage.marginHorizontal,
}))

const Row = styled.View({
  flexDirection: 'row',
  alignItems: 'center',
})

const ContentContainer = styled.View({
  flexDirection: 'row',
  flex: 1,
})

const ButtonContainer = styled.View({
  maxWidth: getSpacing(70),
  flex: 1,
})

const DEFAULT_MARGIN = getSpacing(6)
export const FavoriteButtonsContainer = styled.View(({ theme }) => {
  const WEB_MARGIN_LEFT =
    DEFAULT_MARGIN + theme.tiles.sizes.small.width + getSpacing(SPACER_BETWEEN_IMAGE_AND_CONTENT)
  return {
    flexDirection: 'row',
    marginTop: DEFAULT_MARGIN,
    marginRight: DEFAULT_MARGIN,
    marginLeft: theme.isMobileViewport ? DEFAULT_MARGIN : WEB_MARGIN_LEFT,
  }
})

const Distance = styled(Typo.Body)(({ theme }) => ({
  textAlign: 'right',
  color: theme.colors.greyDark,
}))

const Body = styled(Typo.Body)(({ theme }) => ({
  color: theme.colors.greyDark,
}))

const Separator = styled.View(({ theme }) => ({
  height: 2,
  backgroundColor: theme.colors.greyLight,
  marginHorizontal: getSpacing(6),
  marginVertical: getSpacing(4),
}))
