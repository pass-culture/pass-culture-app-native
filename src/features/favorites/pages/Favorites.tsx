import { t } from '@lingui/macro'
import React from 'react'
import styled from 'styled-components/native'

import { useAuthContext } from 'features/auth/AuthContext'
import { FavoritesResults } from 'features/favorites/components/FavoritesResults'
import { NotConnectedFavorites } from 'features/favorites/components/NotConnectedFavorites'
import SvgPageHeader from 'ui/components/headers/SvgPageHeader'
import { useKeyboardAdjust } from 'ui/components/keyboard/useKeyboardAdjust'

export const Favorites: React.FC = () => {
  useKeyboardAdjust()
  const { isLoggedIn } = useAuthContext()

  if (!isLoggedIn) return <NotConnectedFavorites />

  return (
    <Container>
      <SvgPageHeader title={t`Mes favoris`} />
      <FavoritesResults />
    </Container>
  )
}

const Container = styled.View({ flex: 1 })
