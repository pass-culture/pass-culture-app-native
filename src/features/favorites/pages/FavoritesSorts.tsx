import { t } from '@lingui/macro'
import React, { useState } from 'react'
import { ScrollView, ViewStyle } from 'react-native'
import styled from 'styled-components/native'

import { useFavoritesState } from 'features/favorites/pages/FavoritesWrapper'
import { getTabNavConfig } from 'features/navigation/TabBar/helpers'
import { useGoBack } from 'features/navigation/useGoBack'
import { analytics } from 'libs/analytics'
import { GeolocPermissionState, useGeolocation } from 'libs/geolocation'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { PageHeader } from 'ui/components/headers/PageHeader'
import { InputError } from 'ui/components/inputs/InputError'
import { Validate } from 'ui/svg/icons/Validate'
import { getSpacing, Spacer, Typo } from 'ui/theme'

export type FavoriteSortBy = 'RECENTLY_ADDED' | 'ASCENDING_PRICE' | 'AROUND_ME'

const SORT_OPTIONS: Record<FavoriteSortBy, string> = {
  RECENTLY_ADDED: t`Ajouté récemment`,
  ASCENDING_PRICE: t`Prix croissant`,
  AROUND_ME: t`Proximité géographique`,
}
const SORT_OPTIONS_LIST = Object.entries(SORT_OPTIONS) as Array<[FavoriteSortBy, string]>

export const FavoritesSorts: React.FC = () => {
  const { goBack } = useGoBack(...getTabNavConfig('Favorites'))
  const {
    position,
    positionError,
    permissionState,
    requestGeolocPermission,
    showGeolocPermissionModal,
  } = useGeolocation()
  const { sortBy: selectedSortBy, dispatch } = useFavoritesState()
  const [stagedSelectedSortBy, setStagedSelectedSortBy] = useState(selectedSortBy)

  async function onSortBySelection(sortBy: FavoriteSortBy) {
    function updateSortBySelection() {
      setStagedSelectedSortBy(sortBy)
    }
    if (sortBy === 'AROUND_ME') {
      if (!position && permissionState === GeolocPermissionState.GRANTED) {
        return
      }
      if (position) {
        return updateSortBySelection()
      }
      if (permissionState === GeolocPermissionState.NEVER_ASK_AGAIN) {
        return showGeolocPermissionModal()
      }
      return await requestGeolocPermission({ onAcceptance: updateSortBySelection })
    }
    return updateSortBySelection()
  }

  function onValidation() {
    analytics.logHasAppliedFavoritesSorting({ sortBy: stagedSelectedSortBy })
    dispatch({ type: 'SET_SORT_BY', payload: stagedSelectedSortBy })
    goBack()
  }

  return (
    <Container>
      <PageHeader title={t`Trier`} />
      <ScrollView contentContainerStyle={contentContainerStyle}>
        <Spacer.TopScreen />
        <Spacer.Column numberOfSpaces={16} />

        <TitleContainer>
          <Spacer.Column numberOfSpaces={12} />
          <Spacer.Row numberOfSpaces={6} />
          <Typo.Title4>{t`Trier par`}</Typo.Title4>
        </TitleContainer>

        {SORT_OPTIONS_LIST.map(([sortBy, label]) => {
          const isSelected = stagedSelectedSortBy === sortBy
          return (
            <React.Fragment key={sortBy}>
              <LabelContainer
                key={sortBy}
                onPress={() => onSortBySelection(sortBy)}
                testID={sortBy}>
                <Spacer.Column numberOfSpaces={8} />
                <Spacer.Row numberOfSpaces={6} />
                <Label isSelected={isSelected}>{label}</Label>
                <Spacer.Flex />
                {!!isSelected && <ValidatePrimary />}
              </LabelContainer>
              {sortBy === 'AROUND_ME' && positionError ? (
                <InputError visible messageId={positionError.message} numberOfSpacesTop={1} />
              ) : null}
            </React.Fragment>
          )
        })}
      </ScrollView>

      <ButtonContainer>
        <ButtonPrimary wording={t`Valider`} onPress={onValidation} center />
      </ButtonContainer>
    </Container>
  )
}

const Label = styled(Typo.ButtonText).attrs({
  numberOfLines: 2,
})<{ isSelected: boolean }>(({ isSelected, theme }) => ({
  color: isSelected ? theme.colors.primary : theme.colors.black,
}))

const ValidatePrimary = styled(Validate).attrs(({ theme }) => ({
  color: theme.colors.primary,
  size: theme.icons.sizes.small,
}))``

const Container = styled.View(({ theme }) => ({
  flex: 1,
  backgroundColor: theme.colors.white,
}))

const contentContainerStyle: ViewStyle = {
  flexGrow: 1,
  paddingLeft: getSpacing(2),
  paddingRight: getSpacing(3),
}

const LabelContainer = styled.TouchableOpacity.attrs(({ theme }) => ({
  activeOpacity: theme.activeOpacity,
}))({
  flexDirection: 'row',
  alignItems: 'center',
  marginBottom: getSpacing(4),
})

const TitleContainer = styled.View({
  flexDirection: 'row',
  alignItems: 'center',
  marginBottom: getSpacing(4),
})

const ButtonContainer = styled.View({
  padding: getSpacing(5),
})
