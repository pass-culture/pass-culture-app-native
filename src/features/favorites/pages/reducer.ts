import { FavoriteSortBy } from 'features/favorites/pages/FavoritesSorts'

type FavoritesState = {
  sortBy: FavoriteSortBy
}

export type FavoritesContext = FavoritesState & {
  dispatch: (action: Action) => void
}

export const initialFavoritesState: FavoritesState = {
  sortBy: 'RECENTLY_ADDED',
}

type Action = { type: 'INIT' } | { type: 'SET_SORT_BY'; payload: FavoriteSortBy }

export const favoritesReducer = (state: FavoritesState, action: Action): FavoritesState => {
  switch (action.type) {
    case 'INIT':
      return initialFavoritesState
    case 'SET_SORT_BY':
      return { ...state, sortBy: action.payload }
    default:
      return state
  }
}
