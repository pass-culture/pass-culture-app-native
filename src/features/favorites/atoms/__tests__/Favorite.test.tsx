import { fireEvent, render } from '@testing-library/react-native'
import { rest } from 'msw'
import React from 'react'
import waitForExpect from 'wait-for-expect'

import { navigate } from '__mocks__/@react-navigation/native'
import {
  ExpenseDomain,
  FavoriteCategoryResponse,
  FavoriteResponse,
  UserProfileResponse,
} from 'api/gen'
import { initialFavoritesState } from 'features/favorites/pages/reducer'
import { Credit } from 'features/home/services/useAvailableCredit'
import { env } from 'libs/environment'
import { EmptyResponse } from 'libs/fetch'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { server } from 'tests/server'
import { superFlushWithAct } from 'tests/utils'
import { SNACK_BAR_TIME_OUT } from 'ui/components/snackBar/SnackBarContext'
import { SnackBarHelperSettings } from 'ui/components/snackBar/types'

import { Favorite } from '../Favorite'

const mockShowSuccessSnackBar = jest.fn()
const mockShowErrorSnackBar = jest.fn()
jest.mock('ui/components/snackBar/SnackBarContext', () => ({
  useSnackBarContext: () => ({
    showSuccessSnackBar: jest.fn((props: SnackBarHelperSettings) => mockShowSuccessSnackBar(props)),
    showErrorSnackBar: jest.fn((props: SnackBarHelperSettings) => mockShowErrorSnackBar(props)),
  }),
}))

const credit: Credit = { amount: 100, isExpired: false }
const favorite: FavoriteResponse = {
  id: 393,
  offer: {
    category: {
      categoryType: 'Event',
      label: 'Pratique artistique',
      name: 'LECON',
    } as FavoriteCategoryResponse,
    coordinates: { latitude: 48.9263, longitude: 2.49008 },
    date: null,
    expenseDomains: [ExpenseDomain.All],
    externalTicketOfficeUrl: 'https://externalbooking.test.com',
    id: 146105,
    image: {
      credit: null,
      url:
        'https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CWMA',
    },
    name: 'Un lit sous une rivière',
    price: null,
    startDate: new Date('2021-03-04T20:00:00'),
    startPrice: 270,
    isExhausted: false,
    isExpired: false,
  },
}
const user: UserProfileResponse = {
  isBeneficiary: true,
  bookedOffers: {},
  domainsCredit: { [ExpenseDomain.All]: { initial: 500, remaining: 300 } },
} as UserProfileResponse
const onInAppBooking = jest.fn()

let mockDistance: string | null = null
jest.mock('features/offer/components/useDistance', () => ({
  useDistance: () => mockDistance,
}))

const mockFavoritesState = initialFavoritesState
jest.mock('features/favorites/pages/FavoritesWrapper', () => ({
  useFavoritesState: () => ({
    favoritesState: mockFavoritesState,
  }),
}))

describe('<Favorite /> component', () => {
  afterEach(jest.clearAllMocks)

  it('should navigate to the offer when clicking on the favorite', () => {
    const { getByTestId } = renderFavorite()
    fireEvent.press(getByTestId('favorite'))
    expect(navigate).toHaveBeenCalledWith('Offer', {
      from: 'favorites',
      id: favorite.offer.id,
      shouldDisplayLoginModal: false,
    })
  })

  it('should show distance if geolocation enabled', () => {
    mockDistance = '10 km'
    const { queryByText } = renderFavorite()
    expect(queryByText('10 km')).toBeTruthy()
  })

  it('offer name should take full space if no geolocation', () => {
    mockDistance = '10 km'
    const withDistance = renderFavorite().toJSON()

    mockDistance = null
    const withoutDistance = renderFavorite().toJSON()
    expect(withoutDistance).toMatchDiffSnapshot(withDistance)
  })

  it('should delete favorite on button click', async () => {
    simulateBackend()
    mockDistance = '10 km'
    const { getByText } = renderFavorite()
    await superFlushWithAct()
    fireEvent.press(getByText('Supprimer'))
    await superFlushWithAct()
    await waitForExpect(() => {
      expect(mockShowSuccessSnackBar).toBeCalledWith({
        message: `L'offre a été retirée de tes favoris`,
        timeout: SNACK_BAR_TIME_OUT,
      })
      expect(mockShowErrorSnackBar).not.toBeCalled()
    })
  })
  it('should fail to delete favorite on button click', async () => {
    const id = 0
    simulateBackend({ id, hasRemoveFavoriteError: true })
    mockDistance = '10 km'
    const { getByText } = renderFavorite({
      favorite: { ...favorite, id, offer: { ...favorite.offer, id } },
    })

    await superFlushWithAct(20)
    fireEvent.press(getByText('Supprimer'))
    await superFlushWithAct(20)
    await waitForExpect(() => {
      expect(mockShowSuccessSnackBar).not.toBeCalled()
      expect(mockShowErrorSnackBar).toBeCalledWith({
        message: `L'offre n'a pas été retirée de tes favoris`,
        timeout: SNACK_BAR_TIME_OUT,
      })
    })
  })
})

type Options = {
  id?: number
  hasRemoveFavoriteError?: boolean
}

const DEFAULT_GET_FAVORITE_OPTIONDS = {
  id: favorite.id,
  hasRemoveFavoriteError: false,
}

function simulateBackend(options: Options = DEFAULT_GET_FAVORITE_OPTIONDS) {
  const { id, hasRemoveFavoriteError } = { ...DEFAULT_GET_FAVORITE_OPTIONDS, ...options }
  server.use(
    rest.delete<EmptyResponse>(
      `${env.API_BASE_URL}/native/v1/me/favorites/${id}`,
      (req, res, ctx) =>
        !hasRemoveFavoriteError ? res(ctx.status(204)) : res(ctx.status(422), ctx.json({}))
    )
  )
}

const DEFAULT_PROPS = {
  credit,
  favorite,
  user,
  onInAppBooking,
}

type RenderFavoriteParams = {
  credit?: Credit
  favorite?: FavoriteResponse
  user?: UserProfileResponse
}

function renderFavorite(props: RenderFavoriteParams = DEFAULT_PROPS) {
  const { credit, favorite, user } = { ...DEFAULT_PROPS, ...props }
  return render(
    reactQueryProviderHOC(
      <Favorite credit={credit} favorite={favorite} user={user} onInAppBooking={onInAppBooking} />
    )
  )
}
