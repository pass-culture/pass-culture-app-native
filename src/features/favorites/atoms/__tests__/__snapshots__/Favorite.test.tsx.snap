// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorite /> component offer name should take full space if no geolocation 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -112,10 +112,22 @@
                  \\"flexDirection\\": \\"row\\",
                },
              ]
            }
          >
+           <View
+             flex={0.7}
+             style={
+               Array [
+                 Object {
+                   \\"flexBasis\\": 0,
+                   \\"flexGrow\\": 0.7,
+                   \\"flexShrink\\": 1,
+                 },
+               ]
+             }
+           >
              <Text
                numberOfLines={2}
                style={
                  Array [
                    Object {
@@ -127,10 +139,39 @@
                  ]
                }
              >
                Un lit sous une rivi√®re
              </Text>
+           </View>
+           <View
+             flex={0.3}
+             style={
+               Array [
+                 Object {
+                   \\"flexBasis\\": 0,
+                   \\"flexGrow\\": 0.3,
+                   \\"flexShrink\\": 1,
+                 },
+               ]
+             }
+           >
+             <Text
+               style={
+                 Array [
+                   Object {
+                     \\"color\\": \\"#626262\\",
+                     \\"fontFamily\\": \\"Montserrat-Regular\\",
+                     \\"fontSize\\": 15,
+                     \\"lineHeight\\": 20,
+                     \\"textAlign\\": \\"right\\",
+                   },
+                 ]
+               }
+             >
+               10 km
+             </Text>
+           </View>
          </View>
          <View
            numberOfSpaces={1}
            style={
              Array ["
`;
